### DISCUSSION TITLE
Integrating Adobe Product Recommendations

### DISCUSSION THREAD

**Discussion by Bwilliamson55** [breezefront/community#175] [1. Help] [answered]
Hello! 

I'm fairly new to Breeze, but I'm not that new to Magento.
I've been tinkering with this for a day and a half now- and am wondering what I'm missing here.

Adobe Product Recommendations uses the Adobe Data Service module (Normally) to initiate the events sdk, and collector. 
With Breeze I could follow their 'headless' guide to the letter, but I would prefer to not use CDN packages.
So, I've gone beyond what's necessary, most likely, and redefined the packages needed in a breeze_default.xml:

```
<?xml version="1.0"?>
<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
    <body>
        <referenceBlock name="breeze.js">
            <arguments>
                <argument name="bundles" xsi:type="array">
                    <item name="default" xsi:type="array">
                        <item name="items" xsi:type="array">
                            <item name="recommendationsSDK" xsi:type="string">Mishimoto_ProductRecommendations/js/magento-recs-sdk-v2.js</item>
                            <item name="recommendationsEvents" xsi:type="string">Mishimoto_ProductRecommendations/js/commerce.adobedtm.com_recommendations_events_v1_recommendationsEvents.min.js</item>
                            <item name="dataServicesBase" xsi:type="string">Mishimoto_ProductRecommendations/js/commerce.adobedtm.com_v6_ds.min.js</item>
                            <item name="dataServicesDiscount" xsi:type="string">Mishimoto_ProductRecommendations/js/commerce.adobedtm.com_v6_discount.min.js</item>
                            <item name="magentoStorefrontEvents" xsi:type="string">Mishimoto_ProductRecommendations/js/adobe_magento-storefront-events-sdk1.4.0_dist_index.js</item>
                            <item name="magentoStorefrontEventCollector" xsi:type="string">Mishimoto_ProductRecommendations/js/cdn.jsdelivr.net_npm_@adobe_magento-storefront-event-collector@1_dist_index.js</item>
                        </item>
                    </item>
                </argument>
            </arguments>
        </referenceBlock>
    </body>
</page>
```

Following https://developer.adobe.com/commerce/services/shared-services/storefront-events/sdk/
And https://experienceleague.adobe.com/docs/commerce-merchant-services/product-recommendations/getting-started/headless.html?lang=en
And https://breezefront.com/docs/integration-tips

The primary difficulty is getting the store front events sdk to work from an inline script such as in
The default data service layer : `vendor/magento/module-data-services/view/frontend/layout/default.xml` :
```
    <body>
        <referenceContainer name="head.additional">
            <block class="Magento\DataServices\Block\Context" name="dataservices_base" template="Magento_DataServices::ds.phtml">
                <arguments>
                    <argument name="page_type" xsi:type="string">Default</argument>
                    <argument name="step_name" xsi:type="string">page-view</argument>
                </arguments>
            </block>
        </referenceContainer>
    </body>
```

ie `vendor/magento/module-data-services/view/frontend/templates/ds.phtml` has incompatible inline scripting to set the proper context and fire the page view event:
```
<script>
    require(['magentoStorefrontEvents'], function (magentoStorefrontEvents) {
        if(!magentoStorefrontEvents) return;
...
```
Transforming that to use define lazily :
```
<script data-breeze type="lazy">
    define(['magentoStorefrontEvents'], function(magentoStorefrontEvents) {
```
Does not recognize the module (undefined) even though it is clearly defined in the browser console.

I'm missing something silly here right?

I'd prefer not to re-invent all of the eventing, and I'd much rather convert the existing data service layouts and templates so the recommendation docs are still correct, as in which event types fire on what pages, via the regular layout xmls.

**Comment by vovayatsyuk** [comment]
Hi!

1. No need to replace `require` with `define`.
2. It's expected that `magentoStorefrontEvents` is undefined since breeze takes modules from the $.breezemap object. You have to add the mapping manually.

Add your js file via xml layout update:

```xml
<?xml version="1.0"?>
<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
    <body>
        <referenceBlock name="breeze.js">
            <arguments>
                <argument name="bundles" xsi:type="array">
                    <item name="default" xsi:type="array">
                        <item name="items" xsi:type="array">
                            ...
                            <item name="mishimoto_apr_breezemap" xsi:type="string">Mishimoto_ProductRecommendations/js/breezemap</item>
                        </item>
                    </item>
                </argument>
            </arguments>
        </referenceBlock>
    </body>
</page>
```

Declare your modules in this file:

```js
define([
    'jquery'
], function ($) {
    'use strict';

    $.breezemap.magentoStorefrontEvents = ...;
});
```

p.s. If `magentoStorefrontEvents` is a globally available variable, you can do everything without additional breezemap file:

```html
<script data-breeze type="lazy">
    require(['magentoStorefrontEvents'], function (magentoStorefrontEvents) {
        magentoStorefrontEvents = magentoStorefrontEvents || window.magentoStorefrontEvents;
        //...
    });
</script>
```

**Reply by Bwilliamson55** [reply]
Thanks for the feedback! I'll be testing this soon.

Can you explain the $.breezemap object a little more?

The documentation only has a single mention of it:
![image](https://github.com/breezefront/community/assets/36702082/bb9bbd81-4f58-4ea9-b8f2-287f6df1b9fb)

**Reply by vovayatsyuk** [reply]
`$.breezemap` is an object that is used to retrieve dependencies when `require/define` is called. It's a very simple tool that allows reusing requirejs-based scripts.

For example, when `require(['someDependency'])` is called, [breeze will look](https://github.com/breezefront/module-breeze/blob/master/view/frontend/web/js/core/define.js#L29-L31) if $.breezemap['someDependency'] is available and return it if yes.  Unlike Luma, when dependency is not registered in the $.breezemap, we do not try to load it.

p.s. When you add a `component` prop to the function or object, breeze will register it in $.breezemap automatically.

Example:

```js
// file1.js
var object1 = {
    component: 'myObjectDependency1'
}

// file2.js
var object2 = {...}
$.breezemap['myObjectDependency2'] = object2;

// file3.js
define(['myObjectDependency1', 'myObjectDependency2'], function (dep1, dep2) {
    // dep1 and dep2 will be resolved here!
})
```

**Reply by Bwilliamson55** [reply]
Ok I understand, up until using an external JS package like the Adobe Product Recommendations.

The JS file usually comes from a CDN - <script src="https://unpkg.com/@adobe/magento-storefront-events-sdk/dist/index.js"></script>
From which you can then use 
```
import mse from "@adobe/magento-storefront-events-sdk";

// subscribe to events
mse.subscribe.pageView(pageViewHandler);

// set context data
mse.context.setPage(/* page context */);

// publish events
mse.publish.pageView();
```
Docs: https://developer.adobe.com/commerce/services/shared-services/storefront-events/sdk/ 

Alternatively- could we re-use the existing JS , I think?
Product reccommendations, vendor/magento/module-product-recommendations-layout/view/frontend/requirejs-config.js:
```
var config = {
    shim: {
        recommendationsSDK: {
            exports: "recsSDK",
        },
    },
    paths: {
        recommendationsSDK: "https://magento-recs-sdk.adobe.net/v2/index",
        recommendationsEvents: ['https://commerce.adobedtm.com/recommendations/events/v1/recommendationsEvents.min', 'Magento_ProductRecommendationsLayout/js/noopRecommendationsEvents'],
    },
}
```

How do you define the `component` as an existing file, without modifying it to fit what breeze expects?

**Reply by vovayatsyuk** [reply, marked as answer]
I can't say if you can reuse these without seeing them. Do they expose some entry points after including them on the page?

The one you pointed (https://unpkg.com/@adobe/magento-storefront-events-sdk/dist/index.js) expose itself as `magentoStorefrontEvents` global variable when including without requirejs. I assume the rest files will do the same.

```xml
<item name="items" xsi:type="array">
  <item name="magento-storefront-events" xsi:type="string">.../magento-storefront-events-sdk-1.4.1.js</item>
</item>
```
Now you can use it in any js file:

```js
var mse = window.magentoStorefrontEvents;
```

Or, if you wish to use it in define statement:

```js
// file1.js
$.breezemap['magentoStorefrontEvents'] = window.magentoStorefrontEvents;

//file2.js
define(['magentoStorefrontEvents'], function (magentoEvents) {
  ...
});
```

**Reply by Bwilliamson55** [reply]
Thanks!
That was the bit I wasn't understanding.
1- The name being slugified in the xml (Not sure if that even plays a role)
2- That I can access the libs set in breeze_default.xml as globals

I've got it working now :D
